TH-1437 | Upgrade to node v22, declutter packages/components test output

karisal-anders opened the pull request:
## Description

Upgrade to **node v22** because with node v20 the tests failed and produced a huge amount (~400k rows / 50MB) of test output.

**Declutter packages/components test output**.

See **commits** for **details**.

Only errors remaining when running packages/components tests are instances of:
```
Warning: validateDOMNesting(...): <a> cannot appear as a descendant of <a>.
```
which is a valid error and should be shown.

And `Warning: Each child in a list should have a unique "key" prop.` related to:
> Check the render method of `CarouselSlider`.

in `src/components/domain/event/similarEvents/__tests__/SimilarEvents.test.tsx > similar events > should render similar event cards` which came up after rebasing on origin/main so late in this PR's development, that it won't be addressed in this PR anymore. So to make it clear: this error should be fixed, yes, but it won't be fixed in this PR.

## Related

- [TH-1437](https://helsinkisolutionoffice.atlassian.net/browse/TH-1437)
- Merged DevOps PRs (In cache/dev/review node 20 → 22):
  - [Liikunta](https://dev.azure.com/City-of-Helsinki/liikunta/_git/liikunta-pipelines/pullrequest/12884)
  - [Harrastukset](https://dev.azure.com/City-of-Helsinki/harrastukset/_git/harrastukset-pipelines/pullrequest/12885)
  - [Tapahtumat](https://dev.azure.com/City-of-Helsinki/tapahtumat/_git/tapahtumat-pipelines/pullrequest/12886)

## Testing

### Automated tests

### Manual testing

## Screenshots

## Additional notes


[TH-1437]: https://helsinkisolutionoffice.atlassian.net/browse/TH-1437?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

karisal-anders committed:
feat(tests): enable readable apollo client errors in packages/components

Previously tests showed errors like:
```
An error occurred! For more details, see the full error text at
https://go.apollo.dev/c/err#%7B...
```
where the link went on for many rows.

This enables the apollo client errors messages to be directly readable
rather than having to click on a link and go to a webpage where one can
see the error message.

refs HH-381

karisal-anders committed:
fix(components): make useHdsStyleFix safe to use if there's no document

Noticed that the function was unsafe when running test file
src/app/__tests__/BaseApp.test.tsx tests and got error:
```
ReferenceError: document is not defined
 ❯ Timeout._onTimeout src/app/useHdsStyleFix.tsx:14:20
     12|   React.useEffect(() => {
     13|     setTimeout(() => {
     14|       const body = document?.body;
       |                    ^
     15|
     16|       if (body) {
```

Related documentation:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining
> Optional chaining cannot be used on a non-declared root object, but
> can be used with a root object with value undefined.

with example:

> undeclaredVar?.prop; // ReferenceError: undeclaredVar is not defined

refs HH-381

karisal-anders committed:
feat(common-tests): add hideConsoleMessages function

This can be used anywhere in the monorepo to hide console messages at
specific console levels, e.g. console.log or console.error.

refs TH-1437

karisal-anders committed:
fix(tests): hide sourcemap missing warnings to declutter test output

refs TH-1437

karisal-anders committed:
fix(tests): fix apollo cache writing warnings in PlaceSelector tests

Previously running PlaceSelector.test.tsx gave warnings:
```
Missing field 'en' while writing result { fi: 'Annantalo' }
Missing field 'sv' while writing result { fi: 'Annantalo' }
Missing field 'en' while writing result { fi: 'Annantalo' }
Missing field 'sv' while writing result { fi: 'Annantalo' }
```

Now there are no warnings when running PlaceSelector.test.tsx tests.

refs TH-1437

karisal-anders committed:
fix(tests): fix missing act wrapping warning in MultiSelectDropdown test

Before this running MultiSelectDropdown.test.tsx tests gave warning:
```
Warning: An update to MultiSelectDropdown inside a test was not wrapped
in act(...).

When testing, code that causes React state updates should be wrapped
into act(...):

act(() => {
  /* fire events that update state */
});
/* assert on the output */
```

refs TH-1437

karisal-anders committed:
fix(tests): remove most errors from SearchSelect.test.tsx's output

Before this change running SearchSelect.test.tsx tests would output
over 150 rows of errors message. Now it is reduced to only one warning
thrown by hds-react library:
```
Warning: Li: Support for defaultProps will be removed from function
components in a future major release. Use JavaScript default parameters
instead.
at Li (node_modules\hds-react\cjs\index.js:1:318591)
at Select (packages\components\src\components\select\Select.tsx:11:3)
at SearchSelect (packages\components\src\components\select\SearchSele...
```

For the test cases where error output is expected all error output is
disabled by mocking console.error for those test cases.

refs TH-1437

karisal-anders committed:
fix(tests): hide error message in GeolocationService.test.ts

Now not showing error message, e.g.:
```
17.01.50 [error] [general] - [error] [general] -  Failed to use
geolocation: User denied Geolocation
```
when running the GeolocationService.test.ts tests. Instead hiding the
error message and testing for it.

refs TH-1437

karisal-anders committed:
fix(tests): fix renderHook import in use-deep-compare-memoize.test.ts

Removes the errors/warnings when running use-deep-compare-memoize tests:
```
stderr | src/hooks/__tests__/use-deep-compare-memoize.test.ts >
useDeepCompareMemoize > should not mutate references

Warning: `ReactDOMTestUtils.act` is deprecated in favor of `React.act`.
Import `act` from `react` instead of `react-dom/test-utils`. See
https://react.dev/warnings/react-dom-test-utils for
Warning: ReactDOM.render is no longer supported in React 18. Use
createRoot instead. Until you switch to the new API, your app will
behave as if it's running React 17. Learn more: https://

stderr | src/hooks/__tests__/use-deep-compare-memoize.test.ts >
useDeepCompareMemoize > should not mutate references

Warning: unmountComponentAtNode is deprecated and will be removed in the
next major release. Switch to the createRoot API. Learn more:
https://reactjs.org/link/switch-to-createroot
```

refs TH-1437

karisal-anders committed:
fix(tests): fix warnings in use-promise.test.tsx

Previously these warnings came up with use-promise.test.tsx:
```
Warning: `ReactDOMTestUtils.act` is deprecated in favor of `React.act`.
Import `act` from `react` instead of `react-dom/test-utils`. See
https://react.dev/warnings/react-dom-test-utils for more info.

Warning: ReactDOM.render is no longer supported in React 18. Use
createRoot instead. Until you switch to the new API, your app will
behave as if it's running React 17. Learn more:
https://reactjs.org/link/switch-to-createroot

Warning: unmountComponentAtNode is deprecated and will be removed in the
next major release. Switch to the createRoot API. Learn more:
https://reactjs.org/link/switch-to-createroot
```

Related issue ("No waitForNextUpdate ?"):
https://github.com/testing-library/react-testing-library/issues/1101
where using waitFor instead of waitForNextUpdate is encouraged.

refs TH-1437

karisal-anders committed:
fix(tests): hide & test for console warnings in getSecureImage.test.ts

Now no console warnings are shown when running getSecureImage.test.ts
tests, but they are tested for.

refs TH-1437

karisal-anders committed:
fix: remove use of 10 year old path package, use node.js's path instead

https://www.npmjs.com/package/path was published last 10 years ago:
> This is an exact copy of the NodeJS ’path’ module published to the NPM
> registry.

Switch to using Node.js's own path:
https://nodejs.org/docs/latest-v22.x/api/path.html

This removes deprecation warning:

(node:13560) [DEP0056] DeprecationWarning: The `util.isString` API is
deprecated.  Please use `typeof arg === "string"` instead.
    at Object.win32.join (node_modules\path\path.js:234:15)
    at new NextPageLister (...src\utils\headless-cms\service.tsx:181:32)
...

refs TH-1437

karisal-anders committed:
fix(tests): fix "Missing field providerContactInfo" in EventInfo tests

Make following warning disappear when running EventInfo.test.tsx:
> Missing field 'providerContactInfo' while writing result

by providing null value for providerContactInfo in used mocks.

refs TH-1437

karisal-anders committed:
fix(tests): match mocks in EventContent tests

By matching mocks the warnings about not having mocks available for
made requests are removed → cleaner test run output.

refs TH-1437

karisal-anders committed:
fix(tests): hide error msg caused by HDS 3 Select in packages/components

Hide error message caused by hds-react v3 Select component.

Example use case:
- SearchSelect (packages/components) → Select (packages/components) →
  Select (hds-react)

Related issue:
- https://github.com/facebook/react/issues/29233

Example output from test run:
```
$ yarn test src/components/select/__tests__/SearchSelect.test.tsx

 RUN  v3.2.4 packages/components

stderr | src/components/select/__tests__/SearchSelect.test.tsx >
SearchSelect > successful cases > renders properly

Warning: Li: Support for defaultProps will be removed from function
components in a future major release. Use JavaScript default parameters
instead.
at Li (node_modules\hds-react\cjs\index.js:1:318591)
at Select (packages\components\src\components\select\Select.tsx:11:3)
at SearchSelect (packages\components\src\components\select\SearchSele...
```

refs TH-1437

karisal-anders committed:
fix(tests): fix OtherEventTimes tests, enable skipped test

Reuse mocks many times as they were needed many times in the tests,
tested that reducing the maxUsageCount for the mocks made the mocks not
match anymore. So, yes, some mock is used 20 times and another 7 times.

No more warnings/errors when running OtherEventTimes tests, except for
```
Warning: validateDOMNesting(...): <a> cannot appear as a descendant of
<a>.
```
which is a valid error and should be shown i.e. the components in the
tests are generating invalid HTML and should be fixed.

Also:
 - Move only once used code from separate function straight to the only
   test case using it → code is more in context & passes linter without
   eslint-disable clause

refs TH-1437

karisal-anders committed:
refactor(tests): move EventInfo mocks to mocks directory for clarity

refs TH-1437

karisal-anders committed:
fix(tests): fix EventInfo test case which was incorrectly named

The test's name and implementation were out-of-sync i.e. the test's name
said something different than what was being tested for. Now the test
case's name and implementation are in agreement.

refs TH-1437

karisal-anders committed:
fix(tests): fix EventInfo mocks so they match → declutters test output

refs TH-1437

karisal-anders committed:
fix(tests): enable skipped EventInfo tests & make them pass

Price info was removed from EventInfo component in PR #846:
- https://github.com/City-of-Helsinki/events-helsinki-monorepo/pull/846

Made the ICS file saving tests pass by testing that ics.createEvent is
called, rather than FileSaver.saveAs as testing for the latter wasn't
working for some unknown reason—not ideal, but better than not having
the test cases.

refs TH-1437

karisal-anders committed:
fix(tests): declutter test output in EventsFederationApolloClient tests

Hide e.g.:
```
16.17.20 [info] [graphql-client] - [info] [graphql-client] -  Allowing
unauthorized requests
16.17.20 [info] [graphql-client] - [info] [graphql-client] -  Allowing
unauthorized requests
```
and instead test for "Allowing unauthorized requests" to have been
output to console.info.

refs TH-1437

terovirtanen commented:
**FEDERATION-ROUTER branch is deployed to platta: https://events-graphql-federation-sports-pr862.dev.hel.ninja** :rocket::rocket::rocket:

karisal-anders committed:
feat: upgrade to node v22

This makes packages/components tests pass without a huge amount
(~400k rows / 50MB of test output) of errors/warnings.

Also:
- Remove unused happy-dom package

refs TH-1437

karisal-anders committed:
fix(tests): make use-promise.test.tsx tests more robust with waitFor

Wrap some expects with `await waitFor(() => expect(...))` to add a bit
of robustness, because the following test failed once when running
packages/components tests:
```
FAIL  src/hooks/__tests__/use-promise.test.tsx > usePromise > hook >
should load, return the promise response and rerender
AssertionError: expected null to strictly equal { value: 'cool' }

- Expected:
{
  "value": "cool",
}

+ Received:
null

 ❯ src/hooks/__tests__/use-promise.test.tsx:30:35
     28|       });
     29|
     30|       expect(result.current.data).toStrictEqual(expected);
       |                                   ^
     31|       expect(result.current.isLoading).toStrictEqual(false);
     32|       expect(result.current.error).toBeNull();
```

refs TH-1437

nikomakela reviewed (approved):
Good job.

karisal-anders committed:
chore(events-helsinki): unify next-i18next.config.mjs with other apps

Now the `next-i18next.config.mjs` files are identical for all apps.

refs TH-1437
